File: sub_projs/cllm/core/tools/anthropic-proxy/reference/proxy.py
#!/usr/bin/env python3
"""anthropic-proxy — 讓只會講 OpenAI＋Bearer 的 cllm 打得到 Anthropic。

cllm 硬編 `Authorization: Bearer <key>`＋OpenAI wire format；Anthropic 收 `x-api-key`＋
`anthropic-version`＋Messages 格式。中間擺這支本機轉發代理：cllm 以為在跟 OpenAI 講話，
代理偷偷翻成 Anthropic，回應再翻回去。cllm 的 C++ 核心一行不動。

    python3 proxy.py                    # 監聽 127.0.0.1:8787
    python3 proxy.py --port 9000        # 換埠
    python3 proxy.py --base https://…   # 換上游（預設 https://api.anthropic.com）

代理「無狀態」：key 從 cllm 送來的 Authorization header 當場轉手成 x-api-key，
代理自己不存、不落地任何金鑰。翻譯形狀全在 translate.py（可離線測）。

cllm 端只要把 config 指過來（見 README）：
    endpoint = http://127.0.0.1:8787/v1/chat/completions
    model    = claude-sonnet-5（或別的 Anthropic 模型 id）
    api_key  = sk-ant-...（你的 Anthropic key，照樣填 config，由代理轉手）
"""

from __future__ import annotations

import argparse
import json
import os
import sys
import urllib.error
import urllib.request
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer

import translate

BASE_URL = os.environ.get("ANTHROPIC_BASE_URL", "https://api.anthropic.com")
API_VERSION = os.environ.get("ANTHROPIC_VERSION", "2023-06-01")
MAX_TOKENS = int(os.environ.get("ANTHROPIC_MAX_TOKENS", str(translate.DEFAULT_MAX_TOKENS)))


def _bearer(headers) -> str:
    """從 Authorization: Bearer <key> 掏出 key（cllm 就這一種）。"""
    auth = headers.get("Authorization", "")
    return auth[7:].strip() if auth.startswith("Bearer ") else auth.strip()


def _anthropic_headers(key: str) -> dict:
    return {"content-type": "application/json",
            "x-api-key": key,
            "anthropic-version": API_VERSION}


class Handler(BaseHTTPRequestHandler):
    protocol_version = "HTTP/1.1"  # 要 chunked 串流就得 1.1

    def log_message(self, fmt, *args):  # 靜音預設每請求一行的 stderr 噪音
        pass

    def do_POST(self):
        # cllm 只打一個 endpoint；路徑不強求，指到本代理即視為 chat/completions。
        try:
            length = int(self.headers.get("Content-Length", 0))
            oai = json.loads(self.rfile.read(length) or b"{}")
        except (ValueError, TypeError) as e:
            return self._error(400, "壞掉的請求 JSON：%s" % e)

        key = _bearer(self.headers)
        if not key:
            return self._error(401, "缺 Authorization: Bearer <anthropic-key>——"
                                    "把 sk-ant-... 填進 cllm config 的 api_key。")

        body = translate.to_anthropic(oai, MAX_TOKENS)
        want_stream = bool(body.get("stream"))
        url = BASE_URL.rstrip("/") + "/v1/messages"
        req = urllib.request.Request(url, data=json.dumps(body).encode(),
                                     headers=_anthropic_headers(key), method="POST")
        try:
            if want_stream:
                self._proxy_stream(req)
            else:
                self._proxy_once(req)
        except urllib.error.HTTPError as e:  # 上游 4xx/5xx：翻成 OpenAI 錯誤外殼原樣回
            self._upstream_error(e)
        except urllib.error.URLError as e:
            self._error(502, "連不上上游 %s：%s" % (url, e.reason))

    # ── 非串流：一次收全、翻形狀、回 ──
    def _proxy_once(self, req):
        with urllib.request.urlopen(req) as resp:
            anth = json.loads(resp.read() or b"{}")
        payload = json.dumps(translate.from_anthropic(anth)).encode()
        self.send_response(200)
        self.send_header("Content-Type", "application/json")
        self.send_header("Content-Length", str(len(payload)))
        self.end_headers()
        self.wfile.write(payload)

    # ── 串流：逐筆 Anthropic SSE → 逐筆 OpenAI SSE，chunked 送出 ──
    def _proxy_stream(self, req):
        xlate = translate.StreamXlate()
        self.send_response(200)
        self.send_header("Content-Type", "text/event-stream")
        self.send_header("Cache-Control", "no-cache")
        self.send_header("Transfer-Encoding", "chunked")
        self.end_headers()
        with urllib.request.urlopen(req) as resp:
            for raw in resp:                       # urllib 逐行；Anthropic SSE 是行導向
                line = raw.decode("utf-8", "replace").rstrip("\r\n")
                if not line.startswith("data:"):   # event:／空行／註解都略過
                    continue
                data = line[5:].strip()
                if not data or data == "[DONE]":
                    continue
                try:
                    evt = json.loads(data)
                except ValueError:
                    continue
                for out in xlate.feed(evt):
                    self._chunk(out)
        self._chunk("data: [DONE]\n\n")  # 保底收尾（message_stop 通常已送過，重送無害）
        self._chunk("")                  # 結束 chunked

    # ── HTTP/1.1 chunked 小工具 ──
    def _chunk(self, text: str):
        data = text.encode()
        self.wfile.write(b"%X\r\n%s\r\n" % (len(data), data))
        self.wfile.flush()

    def _error(self, status: int, message: str):
        payload = json.dumps({"error": {"message": message}}).encode()
        self.send_response(status)
        self.send_header("Content-Type", "application/json")
        self.send_header("Content-Length", str(len(payload)))
        self.end_headers()
        self.wfile.write(payload)

    def _upstream_error(self, e: urllib.error.HTTPError):
        try:
            anth_err = json.loads(e.read() or b"{}")
            oai_err = translate.err_to_openai(anth_err)
        except ValueError:
            oai_err = {"error": {"message": "上游錯誤 HTTP %d" % e.code}}
        payload = json.dumps(oai_err).encode()
        self.send_response(e.code)
        self.send_header("Content-Type", "application/json")
        self.send_header("Content-Length", str(len(payload)))
        self.end_headers()
        self.wfile.write(payload)


def main(argv=None) -> int:
    global BASE_URL
    p = argparse.ArgumentParser(prog="anthropic-proxy", description=__doc__.splitlines()[0])
    p.add_argument("--host", default="127.0.0.1")
    p.add_argument("--port", type=int, default=int(os.environ.get("PORT", 8787)))
    p.add_argument("--base", default=BASE_URL, help="上游 base URL（預設 %s）" % BASE_URL)
    args = p.parse_args(argv)

    BASE_URL = args.base
    server = ThreadingHTTPServer((args.host, args.port), Handler)
    print("anthropic-proxy 監聽 http://%s:%d → %s（version %s）" %
          (args.host, args.port, BASE_URL, API_VERSION), file=sys.stderr)
    print("cllm config：endpoint=http://%s:%d/v1/chat/completions、"
          "model=<anthropic 模型 id>、api_key=<sk-ant-...>" %
          (args.host, args.port), file=sys.stderr)
    try:
        server.serve_forever()
    except KeyboardInterrupt:
        print("\n收工。", file=sys.stderr)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: sub_projs/cllm/core/tools/anthropic-proxy/proxy.cpp
// proxy.cpp — anthropic-proxy 執行檔：讓只會講 OpenAI＋Bearer 的 cllm 打得到 Anthropic。
//
// 入站用共用 httpd（tools/common）收 cllm 的 OpenAI 請求；出站重用 cllm 的 http（src/http）
// 打 Anthropic Messages。翻譯全在 translate。無狀態：key 從 cllm 送來的 Authorization 當場
// 轉手成 x-api-key，代理自己不存不落地。
//
//   anthropic-proxy [--host 127.0.0.1] [--port 8787] [--base https://api.anthropic.com]
//   env：ANTHROPIC_BASE_URL / ANTHROPIC_VERSION / ANTHROPIC_MAX_TOKENS / PORT

#include <cctype>
#include <cstdlib>
#include <iostream>
#include <string>

#include "http.hpp"      // cllm 出站笨管子（src/）
#include "httpd.hpp"     // 共用入站 server（tools/common/）
#include "translate.hpp"

namespace {

std::string env_or(const char *k, std::string d) {
  const char *v = std::getenv(k);
  return v && *v ? std::string(v) : d;
}

std::string bearer(const httpd::Request &req) {
  std::string a = req.header("Authorization");
  if (a.rfind("Bearer ", 0) == 0) return a.substr(7);
  return a;
}

std::string dump(axl::J &j) {
  std::string s;
  (void)glz::write_json(j, s);
  return s;
}

std::string err_json(const std::string &msg) {
  return "{\"error\":{\"message\":\"" + msg + "\"}}";
}

// 從 SSE line buffer 抽出完整行，翻譯後 write_chunk；殘料留在 buf。
void drain_sse(std::string &buf, axl::StreamXlate &xlate, httpd::Responder &res) {
  size_t nl;
  while ((nl = buf.find('\n')) != std::string::npos) {
    std::string line = buf.substr(0, nl);
    buf.erase(0, nl + 1);
    if (!line.empty() && line.back() == '\r') line.pop_back();
    if (line.rfind("data:", 0) != 0) continue;
    std::string data = line.substr(5);
    while (!data.empty() && data.front() == ' ') data.erase(data.begin());
    if (data.empty() || data == "[DONE]") continue;
    axl::J evt;
    if (glz::read_json(evt, data)) continue;  // 解不動就跳過
    for (const std::string &out : xlate.feed(evt)) res.write_chunk(out);
  }
}

struct Config {
  std::string base = env_or("ANTHROPIC_BASE_URL", "https://api.anthropic.com");
  std::string version = env_or("ANTHROPIC_VERSION", "2023-06-01");
  int max_tokens = std::atoi(env_or("ANTHROPIC_MAX_TOKENS",
                                    std::to_string(axl::kDefaultMaxTokens).c_str())
                                 .c_str());
};

void handle(const httpd::Request &hreq, httpd::Responder &res, const Config &cfg) {
  if (hreq.method != "POST") {
    res.send_json(405, err_json("只收 POST"));
    return;
  }
  axl::J oai;
  if (glz::read_json(oai, hreq.body) || !oai.is_object()) {
    res.send_json(400, err_json("壞掉的請求 JSON"));
    return;
  }
  std::string key = bearer(hreq);
  if (key.empty()) {
    res.send_json(401, err_json("缺 Authorization: Bearer <anthropic-key>——"
                                "把 sk-ant-... 填進 cllm config 的 api_key。"));
    return;
  }

  axl::J anth = axl::to_anthropic(oai, cfg.max_tokens);
  bool want_stream = anth.is_object() && anth.contains("stream") &&
                     anth["stream"].is_boolean() && anth["stream"].get_boolean();

  http::Request out;
  out.url = cfg.base + "/v1/messages";
  out.method = "POST";
  out.headers = {"content-type: application/json", "x-api-key: " + key,
                 "anthropic-version: " + cfg.version};
  out.body = dump(anth);

  if (!want_stream) {
    http::Response up = http::request(out);
    axl::J parsed;
    bool ok = !glz::read_json(parsed, up.body);
    if (up.status >= 400 || (ok && parsed.is_object() && parsed.contains("error"))) {
      axl::J oe = ok ? axl::err_to_openai(parsed) : axl::J{};
      std::string body = ok ? dump(oe) : err_json("上游錯誤 HTTP " + std::to_string(up.status));
      res.send_json(up.status >= 400 ? up.status : 502, body);
      return;
    }
    axl::J oo = axl::from_anthropic(parsed);
    res.send_json(200, dump(oo));
    return;
  }

  // 串流：Anthropic 上游可能回 SSE（成功）或 JSON 錯誤體（4xx）。先偷看首個非空白字元決定：
  // '{' → 錯誤體（還沒送標頭，可回正常錯誤）；否則 → SSE（begin_chunked 開串）。
  axl::StreamXlate xlate;
  std::string raw, linebuf;
  bool decided = false, is_sse = false;
  int status = http::stream(out, [&](std::string_view part) -> bool {
    raw.append(part);
    if (!decided) {
      size_t i = 0;
      while (i < raw.size() && std::isspace(static_cast<unsigned char>(raw[i]))) ++i;
      if (i >= raw.size()) return false;  // 全空白，再等
      decided = true;
      is_sse = raw[i] != '{';
      if (is_sse) {
        res.begin_chunked(200, "text/event-stream");
        linebuf = raw;  // 已含當前 part
        drain_sse(linebuf, xlate, res);
      }
    } else if (is_sse) {
      linebuf.append(part);
      drain_sse(linebuf, xlate, res);
    }
    return false;
  });

  if (is_sse) {
    res.write_chunk("data: [DONE]\n\n");
    res.end_chunked();
  } else {
    axl::J parsed;
    bool ok = !glz::read_json(parsed, raw);
    axl::J oe = ok ? axl::err_to_openai(parsed) : axl::J{};
    std::string body = ok ? dump(oe) : err_json("上游錯誤 HTTP " + std::to_string(status));
    res.send_json(status >= 400 ? status : 502, body);
  }
}

}  // namespace

int main(int argc, char **argv) {
  std::string host = "127.0.0.1";
  int port = std::atoi(env_or("PORT", "8787").c_str());
  Config cfg;
  for (int i = 1; i < argc; ++i) {
    std::string a = argv[i];
    auto next = [&]() { return i + 1 < argc ? std::string(argv[++i]) : std::string(); };
    if (a == "--host") host = next();
    else if (a == "--port") port = std::atoi(next().c_str());
    else if (a == "--base") cfg.base = next();
    else if (a == "-h" || a == "--help") {
      std::cerr << "用法: anthropic-proxy [--host H] [--port P] [--base URL]\n";
      return 0;
    }
  }

  try {
    httpd::Server server(host, port);
    std::cerr << "anthropic-proxy 監聽 http://" << host << ":" << server.port()
              << " → " << cfg.base << "（version " << cfg.version << "）\n";
    std::cerr << "cllm config：endpoint=http://" << host << ":" << server.port()
              << "/v1/chat/completions、model=<anthropic id>、api_key=<sk-ant-...>\n";
    server.serve_forever(
        [&cfg](const httpd::Request &req, httpd::Responder &res) { handle(req, res, cfg); });
  } catch (const std::exception &e) {
    std::cerr << "錯誤：" << e.what() << "\n";
    return 1;
  }
  return 0;
}
